---
title: "DropClust Clustering"
output: html_notebook
---
# dropClust

[Tutorial](https://github.com/debsin/dropClust)

# Desktop Installation

The developer version of the R package can be installed with the following R commands:

```{r}
library(devtools)
install_github("debsin/dropClust", dependencies = T)
```

# Vignette tutorial

This vignette uses a small data set from the 10X website to demonstrate a standard pipeline. This vignette can be used as a tutorial as well.

## Setting up directories

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
library(dropClust)
library(scater)
set.seed(0)
```

## Loading data

dropClust loads UMI count expression data from three input files. The files follow the same structure as the datasets available from the 10X website, i.e.:

* count matrix file in sparse format
* transcriptome identifiers as a TSV file and
* gene identifiers as a TSV file

```{r}
# Load Data, path contains decompressed files 
sce <-readfiles(path = "C:/Users/asus/Desktop/Silezya_Proje/Proje_Dosyalari/Single-Cell-Project/Methods-of-clustering-single-cell-RNA-sequencing-data/Clustering Methods/hg19/")
```

```{r}
dim(sce)
class(sce)
```

```{r}
# find non-zero counts
nonZero <- (counts(sce) > 0)

# find rows/genes with at least one non-zero count 
keep <- (rowSums(nonZero) > 0)

# keep only the genes with non-zero count the SCE object 
sce_2 <- sce[keep, ]

# explore sce_2
sce_2
```

```{r}
dim(sce_2)
```

For each cell, we calculate quality control metrics such as the total number of counts.
*perCellQCMetrics* function calculates useful QC metrics for identification and removal of potentially problematic cells. Obvious per-cell metrics are the sum of counts (i.e., the library size) and the number of detected features. The percentage of counts in the top features also provides a measure of library complexity.

```{r}
sce_2 <- perCellQCMetrics(sce_2)
par(mfrow=c(1,2))
hist(sce_2$total, xlab="Library sizes", main="",
breaks=20, col="grey80", ylab="Number of cells")
hist(sce_2$detected, xlab="Number of expressed genes", main="",
breaks=20, col="grey80", ylab="Number of cells")
```

## Filtering out low-abundance genes

Low-abundance genes are problematic as zero or near-zero counts do not contain enough information for reliable statistical inference. In addition, the discreteness of the counts may interfere with downstream statistical procedures, e.g., by compromising the accuracy of continuous approximations. Here, low-abundance genes are defined as those with an average count below a filter threshold of 1. These genes are likely to be dominated by drop-out events, which limits their usefulness in later analyses. Removal of these genes mitigates discreteness and reduces the amount of computational work without major loss of information.

```{r}
ave.counts <- rowMeans(counts(sce))
keep <- ave.counts >= 1
sum(keep)
hist(log10(ave.counts), breaks=100, main="", col="grey80",
     xlab=expression(Log[10]~"average count"))
abline(v=log10(0.01), col="blue", lwd=2, lty=2)
```

To check whether the chosen threshold is suitable, we examine the distribution of log-means across all genes. The peak represents the bulk of moderately expressed genes while the rectangular component corresponds to lowly expressed genes. The filter threshold should cut the distribution at some point along the rectangular component to remove the majority of low-abundance genes.

## Pre-processing

dropClust performs pre-processing to remove poor quality cells and genes. dropClust is also equipped to mitigate batch-effects that may be present. The user does not need to provide any information regarding the source of the batch for individual transcriptomes. However, the batch-effect removal step is optional.

Cells are filtered based on the total UMI count in a cell specified by parameter min_count. Poor quality genes are removed based on the minimum number of cells *min_count* with expressions above a given threshold *min_count*.

```{r}
# Filter poor quality cells.  A threshold th corresponds to the total count of a cell.
sce<-FilterCells(sce)
sce<-FilterGenes(sce)
```

## Data normalization and removing poor quality genes

Count normalization is then performed with the good quality genes only. Normalized expression values is computed on the raw count data in a *SingleCellExperiment* object, using the median normalized total count.

```{r}
sce<-CountNormalize(sce)
```

## Selecting highly variable genes

Further gene selection is carried out by ranking the genes based on its dispersion index.

```{r}
# Select Top Dispersed Genes by setting ngenes_keep.
sce<-RankGenes(sce, ngenes_keep = 1000)
```

## Structure Preserving Sampling

Primary clustering is performed in a fast manner to estimate a gross structure of the data. Each of these clusters is then sampled to fine tune the clustering process.

```{r}
sce<-Sampling(sce)
```

## Gene selection based on PCA

Another gene selection is performed to reduce the number of dimensions. PCA is used to identify genes affecting major components.

```{r}
# Find PCA top 200 genes. This may take some time.
sce<-RankPCAGenes(sce)
```

# Clustering

## Fine tuning the clustering process

By default best-fit, Louvain based clusters are returned. However, the user can tune the parameters to produce the desired number of clusters. The un-sampled transcriptomes are assigned cluster identifiers from among those identifiers produced from fine-tuning clustering. The post-hoc assignment can be controlled by setting the confidence value *conf*. High *conf* values will assign cluster identifiers to only those transcriptomes sharing a majority of common nearest neighbours.

```{r}
# When `method = hclust`
# Adjust Minimum cluster size with argument minClusterSize (default = 20)
# Adjust tree cut with argument level deepSplit (default = 3), higher value produces more clusters.
sce<-Cluster(sce, method = "default", conf = 0.8)
```

## Visualizing clusters

Compute 2D embeddings for samples followed by post-hoc clustering.

```{r}
library("SingleCellExperiment", lib.loc="~/R/win-library/4.0")
sce<-PlotEmbedding(sce, embedding = "umap", spread = 10, min_dist = 0.1)
plot_data = data.frame("Y1" = reducedDim(sce,"umap")[,1], Y2 = reducedDim(sce, "umap")[,2], color = sce$ClusterIDs)
```

```{r}
## ScatterPlot(plot_data,title = "Clusters")
ScatterPlot(sce,title = "Clusters")
```

## Find cluster specific Differentially Expressed genes

```{r}
DE_genes_all = FindMarkers(sce, selected_clusters=NA, lfc_th = 1, q_th =0.001, nDE=30)

write.csv(DE_genes_all$genes, 
          file = file.path(tempdir(),"ct_genes.csv"),
          quote = FALSE)
```

## Plot hand picked marker genes

```{r}
marker_genes = c("S100A8", "GNLY", "PF4")

p<-PlotMarkers(sce, marker_genes)
```

## Heat map of top DE genes from each cluster

```{r}
# Draw heatmap
p<-PlotHeatmap(sce, DE_res = DE_genes_all$DE_res,nDE = 10)
print(p)
```
